hs_core/hydroshare/resource.py

```diff
@@ -722,16 +725,32 @@ def delete_resource_file(pk, filename_or_id, user):
 
     return filename_or_id
 
+def get_resource_doi(res_id, pending):
+    doi_str = "http://dx.doi.org/10.4211/hs.{shortkey}".format(shortkey=res_id)
+    if pending:
+        return "{doi}pending".format(doi=doi_str)
+    else:
+        return doi_str
+
+
+def get_activated_doi(doi):
+    idx = doi.find('pending')
+    if idx >= 0:
+        return doi[:idx]
+    else:
+        return doi
 
-def publish_resource(pk):
+def publish_resource(user, pk):
```

Comment:
@hyi - from the user's perspective, at the point they complete the publication process by agreeing to the publication agreement and clicking the "Publish" button, they are done.  There is nothing more they can do.  At the point they click the "Publish" button, they should not be able to edit the resource or make any changes.  It is then up to us to make sure the DOI gets registered.  The congratulatory message should pop up when they click the "Publish" button.  The DOI pending messages in the citation should appear until the DOI is registered (as you have it now).

I think your approach of collecting delinquent DOI registration requests and retrying them again once per day sounds reasonable.  I wouldn't bother at this point with additional messages on the resource landing page, I would just leave the DOI pending message you already have on the landing page there until the DOI is successfully registered. I would, however, build some notification to admins into the DOI registration process so someone can be alerted if things start going badly wrong with Crossref. Experience will tell us if we need to do something more.